pkg/controller/provider-network.go

```diff
@@ -4,126 +4,58 @@ import (
-	if !c.isLeader() {
-		return
-	}
-
-	key, err := cache.MetaNamespaceKeyFunc(obj)
+func (c *Controller) resyncProviderNetworkStatus() {
+	klog.Infof("start to sync ProviderNetwork status")
+	pns, err := c.providerNetworksLister.List(labels.Everything())
 	if err != nil {
-		utilruntime.HandleError(err)
+		klog.Errorf("failed to list provider network: %v", err)
 		return
 	}
 
-	klog.V(3).Infof("enqueue update provider network %s", key)
-	c.updateProviderNetworkQueue.Add(key)
-}
-
-func (c *Controller) runUpdateProviderNetworkWorker() {
-	for c.processNextUpdateProviderNetworkWorkItem() {
-	}
-}
-
-func (c *Controller) processNextUpdateProviderNetworkWorkItem() bool {
-	obj, shutdown := c.updateProviderNetworkQueue.Get()
-	if shutdown {
-		return false
-	}
-
-	err := func(obj interface{}) error {
-		defer c.updateProviderNetworkQueue.Done(obj)
-		var key string
-		var ok bool
-		if key, ok = obj.(string); !ok {
-			c.updateProviderNetworkQueue.Forget(obj)
-			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
-			return nil
-		}
-		if err := c.handleUpdateProviderNetwork(key); err != nil {
-			c.updateProviderNetworkQueue.AddRateLimited(key)
-			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
-		}
-		c.updateProviderNetworkQueue.Forget(obj)
-		return nil
-	}(obj)
-
-	if err != nil {
-		utilruntime.HandleError(err)
-	}
-
-	return true
-}
-
-func (c *Controller) handleUpdateProviderNetwork(key string) error {
-	pn, err := c.providerNetworksLister.Get(key)
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			return nil
-		}
-		return err
-	}
-
 	nodes, err := c.nodesLister.List(labels.Everything())
 	if err != nil {
 		klog.Errorf("failed to list nodes: %v", err)
-		return err
-	}
-
-	if providerNetworkIsReady(pn, nodes) != pn.Status.Ready {
-		newPn := pn.DeepCopy()
-		newPn.Status.Ready = !pn.Status.Ready
-		_, err = c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), newPn, metav1.UpdateOptions{})
-		if err != nil {
-			klog.Errorf("failed to update status of provider network %s: %v", pn.Name, err)
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (c *Controller) resyncProviderNetworkStatus() {
-	nodeList, err := c.nodesLister.List(labels.Everything())
-	if err != nil {
-		klog.Errorf("failed to get nodes %v", err)
-		return
-	}
-	pnList, err := c.providerNetworksLister.List(labels.Everything())
-	if err != nil {
-		klog.Errorf("failed to get provider networks %v", err)
 		return
 	}
 
-	for _, pn := range pnList {
-		if providerNetworkIsReady(pn, nodeList) != pn.Status.Ready {
-			newPn := pn.DeepCopy()
-			newPn.Status.Ready = !newPn.Status.Ready
-			_, err = c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), newPn, metav1.UpdateOptions{})
-			if err != nil {
-				klog.Errorf("failed to update status of provider network %s: %v", pn.Name, err)
+	for _, pn := range pns {
+		var readyNodes, notReadyNodes []string
```

Comment:
The listed results must be treated as read-only, so we should modify the copied object:
```go
for _, cachedPn := range pns {
    pn := cachedPn.DeepCopy();
    // ...
}
```